Ng splitter

anandtiwary opened the pull request:
## Description
Please include a summary of the change, motivation and context.

Adding a splitter component. Instead of using ng-content, this component looks for sibling elements(prev and next) and adjust the width/height according to the movement of the mouse. It supports both horizontal and vertical split.


https://user-images.githubusercontent.com/52081890/105258403-8e64c200-5b3e-11eb-914f-ea8475fd340d.mov

<img width="617" alt="Screen Shot 2021-01-20 at 10 17 59 PM" src="https://user-images.githubusercontent.com/52081890/105291214-6d679580-5b6d-11eb-80fb-f144a50f110d.png">


<!--
- **on a feature**: describe the feature and how this change fits in it, e.g. this PR makes kafka message.max.bytes configurable to better support batching
- **on a refactor**: describe why this is better than previous situation e.g. this PR changes logic for retry on healthchecks to avoid false positives
- **on a bugfix**: link relevant information about the bug (github issue or slack thread) and how this change solves it e.g. this change fixes #99999 by adding a lock on read/write to avoid data races.
-->


### Testing
Please describe the tests that you ran to verify your changes. Please summarize what did you test and what needs to be tested e.g. deployed and tested helm chart locally. 

### Checklist:
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

### Documentation
Make sure that you have documented corresponding changes in this repository or [hypertrace docs repo](https://github.com/hypertrace/hypertrace-docs-website) if required.

<!--
Include __important__ links regarding the implementation of this PR.
This usually includes and RFC or an aggregation of issues and/or individual conversations that helped put this solution together. This helps ensure there is a good aggregation of resources regarding the implementation.
-->

anandtiwary commented:
Worked on it during holidays. Will add tests post first round of feedback.

anandtiwary commented on projects/components/src/splitter/splitter.service.ts:
```diff
@@ -0,0 +1,106 @@
+    return (
+      (direction === SplitterDirection.Horizontal ? parentBounds.width : parentBounds.height) -
+      previousElementUpdatedDimension -
+      4
```
This is the splitter dom's width/height based on direction. This should be an input param .

aaron-steinfeld commented on projects/components/src/public-api.ts:
```diff
@@ -186,6 +186,11 @@ export * from './select/select-option.component';
 export * from './select/select.module';
 
+// Splitter
+export * from './splitter/splitter';
```
Do we have a use case for this?

aaron-steinfeld replied:
sorry meant to delete this when I saw it later :)

anandtiwary replied:
No prob. Waterfall is a strong use case for this. Another use case is the upcoming geo map.  

It also think we can use it at few places where we use the two panel approach like explorer. Splitter would smooth out the interaction there.

aaron-steinfeld commented on projects/components/src/splitter/splitter.component.ts:
```diff
@@ -0,0 +1,68 @@
+        .pipe(
+          debounceTime(this.debounceTime),
+          tap(layoutChange => this.layoutChange.emit(layoutChange)),
+          tap(layoutChange => layoutChange && this.layoutChangeService.publishLayoutChange())
```
rather than using a tap operator, it looks like these actions on layout change belong in the subscriber (which is currently empty)

anandtiwary replied:
sure.

aaron-steinfeld commented on projects/distributed-tracing/src/shared/dashboard/widgets/waterfall/waterfall/waterfall-chart.component.scss:
```diff
@@ -17,16 +17,16 @@
-    min-width: 500px;
-    flex: 1 1;
+    width: 500px;
+    //flex: 1 1;
```
remove commented out

anandtiwary replied:
yep. Will remove it with tests

aaron-steinfeld commented on projects/components/src/splitter/splitter.service.ts:
```diff
@@ -0,0 +1,104 @@
+    let layoutChanged = false;
+    data.forEach(datum => {
+      if (direction === SplitterDirection.Horizontal) {
+        this.updateWidth(datum.element, datum.dimension);
```
this seems inverted to me. I'd expect a horizontal splitter updates height, and a vertical width. In other words, the orientation describes the splitter itself, not how it affects the dimensions of its siblings.

anandtiwary replied:
Okay, so i added a `direction` property which represents the direction in which the splitter could move. It does not represent the orientation of splitter. If this is causing confusion then I could rename the property.

aaron-steinfeld replied:
It's the same purpose though right? When describing a splitter as vertical, I'd expect it to render... vertically. Kind of example: https://material.angular.io/components/divider/overview#vertical-divider

anandtiwary replied:
Yeah, my take was that the direction it can move in is more important. But i can see it is confusion. Will update.

aaron-steinfeld commented on projects/components/src/splitter/splitter.service.ts:
```diff
@@ -0,0 +1,104 @@
+  }
+
+  private maybeUpdateLayoutOfSiblings(data: UpdatedDimensionData[], direction: SplitterDirection): boolean {
+    let layoutChanged = false;
```
this doesn't really seem necessary, it's just being set if there's any data, so we could just return `data.length > 0`

anandtiwary replied:
Oh yeah. I thought such a return would look weird as it didn't depend on any work that happened in the method. I can change it.

anandtiwary replied:
removed

aaron-steinfeld commented on projects/components/src/splitter/splitter.service.ts:
```diff
@@ -0,0 +1,104 @@
+  }
+
+  private updateWidth(element: HTMLElement, width: number): void {
+    element.style.width = `${width}px`;
```
in general (for all the dom manipulation/reading) can we use angular's renderer class?

anandtiwary replied:
I can check that. would it bring this change inside the scope of Change detection?

aaron-steinfeld replied:
I don't think it would affect change detection (since it's not modifying inputs), and it may not give us any direct benefits right here, but it's a good practice for some reasons that are relevant and some that aren't. In descending order of importance:
1. Angular will do any sanitization (e.g. xss) that we'd normally get in a template but not when we directly modify the DOM.
2. With all changes going through a common renderer, we could in the future choose to link that up to a service like layout change (because we can insert custom renderers).
3. If we ever choose to use something like web workers or server side rendering, direct DOM access would break (while this abstraction would work).

anandtiwary replied:
okay, let me check

anandtiwary replied:
I don't see any method to access previous sibling using angular renderer2. I am not sure this would work with the current approach.

aaron-steinfeld replied:
Yeah, perhaps not. For what it's worth - the approach seems very un-angulary and has some other pitfalls due to that too. It's reaching into the dom outside it and manipulating it, which is broken encapsulation and also makes assumptions on how those things are laid out (for example, if one sibling was using flex, this would break, right?)

The API I'd expect for something like this would be:
```
<ht-splitter [direction]="direction">
   <div *htSplitterContent>Left</div> 
   <div *htSplitterContent>Right</div>
</ht-splitter
```
This would keep everything internal to the splitter.

anandtiwary replied:
Nope. it woudn't break if it is inside flex container. We could always nest it in various components and directives, but that just add additional structures to DOM. It doesn't do anything else apart from that. Take a deeply nested splitter as an example.

aaron-steinfeld replied:
It would nest just fine in this example. It does add a dom layer to ensure we're modifying something we own rather than breaking encapsulation, but that seems like a tradeoff I'd take any day.

anandtiwary replied:
Done

anandtiwary replied:
Directive and Angular Renderer

jake-bassett commented:
This would have been a good opportunity to start bringing in [PrimeNG](https://www.primefaces.org/primeng/showcase/#/splitter).

anandtiwary commented:
> This would have been a good opportunity to start bringing in [PrimeNG](https://www.primefaces.org/primeng/showcase/#/splitter).

Do we really want to bring in a new third party lib for small components? We already have a vast collection of in house components, switching to a third party seems a little late now :P.

jake-bassett commented:
> > This would have been a good opportunity to start bringing in [PrimeNG](https://www.primefaces.org/primeng/showcase/#/splitter).
> 
> Do we really want to bring in a new third party lib for small components? We already have a vast collection of in house components, switching to a third party seems a little late now :P.

I think it still makes sense since we are still building and improving components. That component lib brings in a lot more that we still need to build out beyond just the splitter.

anandtiwary commented:
Addressed previous comments

itssharmasandeep commented on projects/components/src/public-api.ts:
```diff
@@ -296,6 +296,16 @@ export * from './select/select-control-option.component';
+export * from './splitter/splitter.component';
+export * from './splitter/splitter.module';
+
+// Multi-select
+export * from './multi-select/multi-select.component';
+export * from './multi-select/multi-select.module';
```
why this is needed, I guess it's not.

anandtiwary replied:
it should be exported already. Let me remove it.

itssharmasandeep commented on projects/components/src/splitter/splitter.component.scss:
```diff
@@ -0,0 +1,81 @@
+@import 'mixins';
+
+:host {
+  width: 100%;
+  height: 100%;
```
shouldn't require host properties

anandtiwary replied:
Yeah, i forgot to remove it.

itssharmasandeep commented on projects/components/src/splitter/splitter.component.scss:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  .splitter-content > p {
+    width: 100%;
+    height: 100%;
```
nit: can be replaced with `fill-container` and all the other places as well

anandtiwary replied:
done

itssharmasandeep commented on projects/components/src/splitter/splitter.component.scss:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  .splitter {
+    display: flex;
+    align-items: center;
+    justify-content: center;
```
nit: can be replaced with `center-contents` mixin

anandtiwary replied:
I prefer to type it out for better readability.

itssharmasandeep commented on projects/components/src/splitter/splitter.component.scss:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    &.vertical {
+      width: calc(100% - 72px);
```
shouldn't require calc IMO, should find a different way instead

jake-bassett replied:
calc() offsets are often a sign that a flexbox is using 100% height incorrectly.

anandtiwary commented:
> > > This would have been a good opportunity to start bringing in [PrimeNG](https://www.primefaces.org/primeng/showcase/#/splitter).
> > 
> > 
> > Do we really want to bring in a new third party lib for small components? We already have a vast collection of in house components, switching to a third party seems a little late now :P.
> 
> I think it still makes sense since we are still building and improving components. That component lib brings in a lot more that we still need to build out beyond just the splitter.

mine is better ;)

(NgPrime splitter does not support dynamic content )

itssharmasandeep commented on projects/components/src/splitter/splitter.component.ts:
```diff
@@ -0,0 +1,294 @@
+    </div>
+  `
+})
+export class SplitterComponent implements OnChanges, AfterContentInit {
```
nit: Please use `protected` instead of `public`

jake-bassett replied:
Hah, this is one that @anandtiwary got me on recently! ;-)

anandtiwary replied:
done

anandtiwary replied:
I did do it for a few properties :)

itssharmasandeep commented on projects/components/src/splitter/splitter.module.ts:
```diff
@@ -0,0 +1,16 @@
+import { MemoizeModule } from '@hypertrace/common';
+
+@NgModule({
+  imports: [CommonModule, IconModule, PopoverModule, ButtonModule, TooltipModule, MemoizeModule],
```
do we need `PopoverModule`?

anandtiwary replied:
no.

itssharmasandeep commented on projects/components/src/splitter/splitter.component.ts:
```diff
@@ -0,0 +1,294 @@
+
+        <div
+          class="splitter"
+          *ngIf="
```
shouldn't the ngIf on `this.direction` should be on the parent div?

anandtiwary replied:
Let me set a default value and move the ngIf

jake-bassett reviewed (approved):
A few nits remaining, but @anandtiwary said he will return to those in a follow up. lgtm otherwise.

anandtiwary commented:
Merging this PR after more than 2 years. :)

anandtiwary commented:
> A few nits remaining, but @anandtiwary said he will return to those in a follow up. lgtm otherwise.

I think I addressed all the review comments. There are a few minor resize bugs, that we will work on in a follow up.
